ipa-core/src/protocol/dp/distributions.rs

```diff
@@ -125,4 +166,116 @@ mod tests {
+        let mut samples = Vec::new();
+        // Sample number_samples values from the generate_truncated_double_geometric function
+        for _ in 0..number_samples {
+            let sample = generate_truncated_double_geometric(s, n, &mut rng);
+            samples.push(sample);
+        }
+        // Compute the sample mean
+        let sample_mean = samples.iter().sum::<isize>() as f64 / samples.len() as f64;
+        // println!("sample_mean: {:?}", sample_mean);
+        // Check that the sample mean is within some distance of the expected value
+        let expected_mean = n as f64;
+        // println!("expected_mean: {:?}", expected_mean);
+        (sample_mean >= expected_mean - t) && (sample_mean <= expected_mean + t)
+    }
+    #[test]
+    fn test_generate_truncated_double_geometric_sample_dist() {
+        let mut rng = rand::thread_rng();
+        let epsilon = 1.0;
+        let s = 1.0 / epsilon;
+        let n = 25;
+        let num_samples = 100_000;
+        let mut samples = Vec::new();
+        // Sample 1000 values from the generate_truncated_double_geometric function
+        for _ in 0..num_samples {
+            let sample = generate_truncated_double_geometric(s, n, &mut rng);
+            assert!(sample >= 0 && sample <= (2 * n));
+            samples.push(sample);
+        }
+        // Compute the observed probability for each value in the range [0, 2*n)
+        let mut histogram = HashMap::new();
+        for value in samples {
+            *histogram.entry(value).or_insert(0) += 1;
+        }
+        let mut sorted_keys: Vec<isize> = histogram.keys().copied().collect();
+        sorted_keys.sort_unstable();
+        // Compute the expected probability for each value in the range [0, 2*n]
+        let normalizing_factor = (1.0 - E.powf(-epsilon))
+            / (1.0 + E.powf(-epsilon) - 2.0 * E.powf(-epsilon * ((n + 1) as f64))); // 'A' in paper
+
+        for x in 0..=(2 * n) {
+            // Compare the observed and expected probabilities for each value in the range [0, 2*n]
+            let observed_probability = histogram
+                .get(&x)
+                .map_or(0.0, |count| f64::from(*count) / f64::from(num_samples));
+            let expected_probability =
+                normalizing_factor * E.powf(-epsilon * ((n - x).abs() as f64));
+            // println!("x, prob: {}, {}",x,expected_probability);
+            // println!("Value: {}, Observed Probability: {:.4}, Expected Probability: {:.4}", x, observed_probability, expected_probability);
+            assert!(
+                (observed_probability - expected_probability).abs() <= 0.01,
+                "Observed probability is not within 1% of expected probability"
+            );
+        }
+    }
+    #[test]
+    fn test_truncated_double_geometric() {
+        let mut rng = rand::thread_rng();
+        let distribution = TruncatedDoubleGeometric { s: 1.0, shift: 25 };
+        distribution.sample(&mut rng);
+    }
```

Comment:
you should probably test that invalid values for shift/probabilities are rejected

Reply:
added more tests like that